### PR TITLE
[backport] Fix dtypes in `cupy.linalg`

### PR DESCRIPTION
**Pull request by emcastillo**
Backport of #4363

There were several conflicts so I expect this to need additional work.

### CODE REVIEW

**Comment by emcastillo** [comment]
[automatic post] Jenkins, test this please.

**Comment by emcastillo** [comment]
Jenkins, test this please

File: cupy/linalg/_solve.py [hunk 1]
```diff
@@ -301,22 +298,19 @@ def inv(a):
...
+    dtype, out_dtype = _util.linalg_common_type(a)
+
+    order = 'F' if a._f_contiguous else 'C'
+    # prevent 'a' to be overwritten
+    a = a.astype(dtype, copy=True, order=order)
 
     cusolver_handle = device.get_cusolver_handle()
     dev_info = cupy.empty(1, dtype=numpy.int32)
-
+    print('types ', a.dtype, dtype, out_dtype)
```
Comment (toslunar) [inline cupy/linalg/_solve.py:313]:
```suggestion
```

**Review by toslunar — CHANGES_REQUESTED** [review]
LGTM except for the line.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Review by toslunar — APPROVED** [review]

**Pull request merged**